### PR TITLE
Implement punycode encoding/decoding

### PR DESCRIPTION
**Pull request by rschu1ze**
Inspired by #34439, especially https://github.com/ClickHouse/ClickHouse/pull/34439#issuecomment-1856806184, which was dormant for too long. This PR also uses the ada-idna library instead of a 15 year old homegrown implementation.

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added functions for punycode encoding/decoding: `punycodeEncode()` and `punycodeDecode()`.

### Documentation entry for user-facing changes
- [ X ] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 19c8ac567f2e92c46d2a3ef86d9d293288143d28 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/57969/19c8ac567f2e92c46d2a3ef86d9d293288143d28/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/7239887222">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/57969/19c8ac567f2e92c46d2a3ef86d9d293288143d28/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by alexey-milovidov** [comment]
Seeing @lemire as one of the main authors of the library gives me full confidence!

File: src/Functions/punycode.cpp [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+            ada::idna::utf8_to_utf32(value, value_length, value_utf32.data());
+
+            bool ok = ada::idna::utf32_to_punycode(value_utf32, value_puny);
+            if (!ok)
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "Internal error during Punycode encoding");
+
+            const size_t bytes_to_write = value_puny.size() + 1;
+            if (res_data_bytes_written + bytes_to_write > res_data.size())
+            {
+                size_t new_size = std::max(res_data.size() * 2, res_data_bytes_written + bytes_to_write);
```
Comment (vitlibar) [inline src/Functions/punycode.cpp:61]: The type of `res_data` is `ColumnString::Chars`, i.e. `PaddedPODArray<UInt8>`, and `PODArray` already can allocate memory by powers of `2`, and also `PODArray` can copy memory, so both `std::max(res_data.size() * 2` and `std::memcpy()` are not so needed and this code can be simplified:

```
res_data.insert(value_puny.c_str(), value_puny.c_str() + value_puny.length() + 1);
```

File: src/Functions/punycode.cpp [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+struct PunycodeEncodeImpl
+{
+    static void vector(
+        const ColumnString::Chars & data,
+        const ColumnString::Offsets & offsets,
+        ColumnString::Chars & res_data,
+        ColumnString::Offsets & res_offsets)
+    {
+        const size_t rows = offsets.size();
+        res_data.resize(rows * 64); /// just a guess
```
Comment (vitlibar) [inline src/Functions/punycode.cpp:36]: `res_data.reserve(data.size())` seems to be a better guess (like we can preallocate memory assuming that `data` contains only Latin characters).

File: src/Functions/punycode.cpp [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+            res_data[res_data_bytes_written] = '\0';
+            res_data_bytes_written += 1;
+
+            res_offsets[row] = prev_res_offset + bytes_to_write;
+
+            prev_offset = offsets[row];
+            prev_res_offset = res_offsets[row];
+            value_utf32.clear();
+            value_puny.clear();
```
Comment (vitlibar) [inline src/Functions/punycode.cpp:76]: It is not really obvious that `ada::idna::utf32_to_punycode()` and the other function requires the output parameter to be empty. I think a comment here would be useful.

File: src/Functions/punycode.cpp [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+{
+    static void vector(
+        const ColumnString::Chars & data,
+        const ColumnString::Offsets & offsets,
+        ColumnString::Chars & res_data,
+        ColumnString::Offsets & res_offsets)
+    {
+        const size_t rows = offsets.size();
+        res_data.resize(rows * 64); /// just a guess
+        res_offsets.resize(rows);
```
Comment (vitlibar) [inline src/Functions/punycode.cpp:37]: `res_data` and `res_offset` both have `reserve()` functions, i.e. they can preallocate memory. So we don't really have to do that here manually and we can simplify the code.

**Comment by lemire** [comment]
> Seeing @lemire as one of the main authors of the library gives me full confidence!

We stand behind this library 100%. It is being used by millions of people as it is part of Node.js. We've had very few bug reports. We are going to keep improving it and extending it.

**Comment by anonrig** [comment]
Good to see Clickhouse adopting Ada Idna! Happy to help with any issues you might encounter.

File: src/Functions/punycode.cpp (hunk 4 above)
Comment (rschu1ze) [reply src/Functions/punycode.cpp:37]: All manual memory management is now gone. `prev_res_offset` is also gone. But I don't see a prettier alternative to `prev_offset` (--> kept it).

**Review by vitlibar — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
It is way too low-level to be user-friendly: https://pastila.nl/?0196792d/9db06e3e231f4096cc5ef0105a936dfa#cVjN8dNOdSrP22FyyT+pvw==
I'll revert for now, let's resubmit.

**Comment by alexey-milovidov** [comment]
Also, see how the error message is wrong. It shows "Internal error" and nothing else. It could not be alright.

**Comment by alexey-milovidov** [comment]
Or keep these functions around and introduce new ones - `IDNAEncode`, `IDNADecode`?

**Comment by azat** [comment]
>If the name is not decodable, we should return the default value (empty string), because otherwise the function is difficult to use on large real datasets.

Sometimes such functions can be used to verify that the domain is validly encoded domain, though I guess sometimes it make sense to return default value. So maybe both variants should be added or something like `OrNull` equivalent.

**Comment by vitlibar** [comment]
> Or keep these functions around and introduce new ones - `IDNAEncode`, `IDNADecode`?

I think it would be better to have both functions: `punycodeEncode`, `punycodeDecode` and `idnaEncode`, `idnaDecode`:
```
> SELECT punycodeEncode('München.München-Ost')
Mnchen.Mnchen-Ost-22bh

> SELECT idnaEncode('München.München-Ost')
xn--mnchen-3ya.xn--mnchen-ost-9db
```

**Comment by alexey-milovidov** [comment]
+1, agree - let's have both functions.

**Comment by lemire** [comment]
@alexey-milovidov ClickHouse is fantastic. :-)

**Pull request merged**